### ISSUE TITLE
"kubectl rollout history --revision=n" produces wrong/inconsistent output with "-o yaml"?

### ISSUE THREAD

**Issue by waynesi** [kubernetes/kubectl#598]
**Is this a request for help?** (If yes, you should use our troubleshooting guide and community support channels, see http://kubernetes.io/docs/troubleshooting/.): No

**What keywords did you search in Kubernetes issues before filing this one?** (If you have found any duplicates, you should instead reply there.): kubectl rollout

---

**Is this a BUG REPORT or FEATURE REQUEST?** (choose one): Bug Report

**Kubernetes version** (use `kubectl version`):
```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.4", GitCommit:"c27b913fddd1a6c480c229191a087698aa92f0b1", GitTreeState:"clean", BuildDate:"201
9-02-28T13:37:52Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.1", GitCommit:"eec55b9ba98609a46fee712359c7b5b365bdd920", GitTreeState:"clean", BuildDate:"201
8-12-13T10:31:33Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Cloud provider or hardware configuration**: GCP n1-highcpu-2 (2 vCPUs, 1.8 GB memory)
- **OS** (e.g. from /etc/os-release):  Ubuntu 16.04.6 LTS (Xenial Xerus)
- **Kernel** (e.g. `uname -a`):  `` Linux node0 4.15.0-1027-gcp #28~16.04.1-Ubuntu SMP Fri Jan 18 10:10:51 UTC 2019 x86_64 x86_64 x86_64 GNU/Linux``
- **Install tools**: kubeadm

```
kubeadm version: &version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.4", GitCommit:"c27b913fddd1a6c480c229191a087698aa92f0b1", GitTreeState:"clean", BuildDate:"2
019-02-28T13:35:32Z", GoVersion:"go1.11.5", Compiler:"gc", Platform:"linux/amd64"}
```

- **Others**:

**What happened**:
I was learning the basic of rolling update on a DaemonSet. The template is very simple ...
```
apiVersion: extensions/v1beta1
kind: DaemonSet
metadata:
  name: ds-one
spec:
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
        system: DaemonSetOne
    spec:
      containers:
      - name: nginx
        image: nginx:1.9.1
        ports:
        - containerPort: 80
```

After I used ``kubectl set image ds ds-one nginx=nginx:1.12.1-alpine`` to flip the image between nginx:1.9.1 and 1.12.1-alpine back and forth a few times (and deleted the pods to get them updated), I run ``kubectl rollout history daemonset ds-one`` to check the rollout history ...

```
daemonset.extensions/ds-one 
REVISION  CHANGE-CAUSE
3         <none>
4         <none>
```

Then I use ``kubectl rollout history daemonset ds-one --revision=3`` and ``... --revision=4`` to check the details of each revision.
```
daemonset.extensions/ds-one with revision #3
Pod Template:
  Labels:       app=nginx
        system=DaemonSetOne
  Containers:
   nginx:
    Image:      nginx:1.9.1
    Port:       80/TCP
    Host Port:  0/TCP
    Environment:        <none>
    Mounts:     <none>
  Volumes:      <none>
```

```
daemonset.extensions/ds-one with revision #4
Pod Template:
  Labels:       app=nginx
        system=DaemonSetOne
  Containers:
   nginx:
    Image:      nginx:1.12.1-alpine
    Port:       80/TCP
    Host Port:  0/TCP
    Environment:        <none>
    Mounts:     <none>
  Volumes:      <none>
```

However, when I repeated the same  two commands with extra "-o yaml", I now got the exact same results which say ``- image: nginx:1.12.1-alpine`` (the latest revision) regardless which revision I specified in the command.

**What you expected to happen**:
The help says ``-o`` only applies to the output format while ``--revision`` shows the details. So using ``-o`` together with ``--revision`` should not change the details produced by ``--revision`` option I reckon?

**Comment by stefaneg** [comment]
Encountered this issue right now. Makes very hard to rollback precisely to old revisions if you cannot see what image it is referring to.

**Comment by waynesi** [comment]
Thanks @stefaneg for your follow-up. I was almost convinced my bug report was talking rubbish as nobody   gave it a damn!

**Comment by stefaneg** [comment]
Interestingly it gave me the right image info if the -o yaml option is omitted. Which is probably the reason why people do not care more about this.

```
kubectl rollout history deployment.v1.apps/myown-deployment --revision=182 | grep Image
```

**Comment by waynesi** [comment]
Hi @stefaneg , maybe my bug report was poorly written then. I did mean the command gave me the right image version info without `-o yaml` - 
* With `-o yaml`, I got the same version number regardless which revision I was checking.
* Without `-o yaml`, I got the correct version number corresponding to the revision I was checking.

**Comment by weixiao619** [comment]
Is this problem solved？

**Comment by maxweiss-74656** [comment]
I am noticing this behavior with daemonsets and -o yaml as well on 1.20.9

**Comment by fghso** [comment]
Same here with 1.20.7:
- With **-o yaml**, I got the same version number regardless which revision I was checking.
- Without **-o yaml**, I got the correct version number corresponding to the revision I was checking.

**Comment by waynesi** [comment]
Issue reopened as it seems (finally) gathering some interest from other people.

**Comment by afreeland** [comment]
I believe I have been seeing this issue was well.  When using the `-o yaml` I seem to no longer get it for the revision requested but rather for the current revision.  Anyone able to get past this?  Had some spikes in a dev environment and wanted to see what commit started it...

**Comment by waynesi** [comment]
Man, I don't know what to do with this ticket tbh. Every time it got some interest immediately after the bot auto-closed this. The bot says "The Kubernetes project currently lacks enough active contributors", so I reckon it means no chance to get it solved unless one would like to contribute. I'm not a Golang dev and I don't understand Kubernetes internal unfortunately.

**Comment by maxweiss-74656** [comment]
> Man, I don't know what to do with this ticket tbh. Every time it got some interest immediately after the bot auto-closed this. The bot says "The Kubernetes project currently lacks enough active contributors", so I reckon it means no chance to get it solved unless one would like to contribute. I'm not a Golang dev and I don't understand Kubernetes internal unfortunately.

Same.  I tried to find it in the code a few weeks ago, and I couldn't. I just don't know Go.

**Comment by afreeland** [comment]
Okay, good to know I am at least not crazy lol. I kept walking through each revision with the `-o yaml` flag and looking at the `creationTimeStamp` and `image` and was like that is weird...why did we do the same thing so many times.  Then I noticed the revision was staying the exact same hah.  

If I can get some free time I can try to poke around and see if I can find anything. Any good places to start or helpful links to jump start it?

**Comment by brianpursley** [comment]
If it is still an issue we can keep it open.  Make sure you 👍 the issue as that helps bring attention to it.  

It should at least get triaged.  I'll see if this can be added to the list for the next bub review 

/reopen

**Comment by waynesi** [comment]
Guys I reopened this ticket again. Hope some magic may happen this time.

**Comment by tomasz-polgrabia** [comment]
This bug I discovered too while taking CKAD exam :-). Sounds like I have perfect timing as always.

**Comment by eddiezane** [comment]
/assign @soltysh 

Is this the same as https://github.com/kubernetes/kubectl/issues/1137?

**Comment by waynesi** [comment]
Not sure to whom this question is. But I had a look at 1137 and didn't feel it's anything related; at least not by the commands and options involved.

**Comment by vladimir259** [comment]
> Same here with 1.20.7:
> 
>     * With **-o yaml**, I got the same version number (the most recent) regardless which revision I was checking.
> 
>     * Without **-o yaml**, I got the correct version number corresponding to the revision I was checking.

confirming, still an issue.

k8s 1.18
kubectl 1.23

**Comment by lawschlosser** [comment]
From what I can tell this issue isn't just about using `-o yaml`.   It's about using the `-o` / `--output` flag _at all_, i.e. `-o json` exhibits the same bug (it did output json though :smiley: )
```
kubectl version
Client Version: version.Info{Major:"1", Minor:"22", GitVersion:"v1.22.3", GitCommit:"c92036820499fedefec0f847e2054d824aea6cd1", GitTreeState:"clean", BuildDate:"2021-10-27T18:41:28Z", GoVersion:"go1.16.9", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"18+", GitVersion:"v1.18.20-gke.901", GitCommit:"db8463aecdfe0eb5d9067effa2fd5ab3ff7a988e", GitTreeState:"clean", BuildDate:"2021-08-04T23:27:42Z", GoVersion:"go1.13.15b4", Compiler:"gc", Platform:"linux/amd64"}
```

**Comment by Sharsie** [comment]
Still an issue with k8s v1.21.9, kubectl v1.24.1

/remove-lifecycle rotten

**Comment by brianpursley** [comment]
I believe this is the same issue as https://github.com/kubernetes/kubernetes/issues/110097, which was recently fixed by https://github.com/kubernetes/kubernetes/pull/111093 which will be included in kubectl 1.26.

Before fix (notice wrong image is returned, `busybox:1.35`, illustrating the bug):
```
~/go/src/k8s.io/kubernetes (master) $ kubectl rollout history deploy/foo --revision=3 | grep busybox
    Image:	busybox:1.32
~/go/src/k8s.io/kubernetes (master) $ kubectl rollout history deploy/foo --revision=3 -o yaml | grep "image: busybox"
        image: busybox:1.35
```

After fix:
```
~/go/src/k8s.io/kubernetes (master) $ _output/bin/kubectl rollout history deploy/foo --revision=3 | grep busybox
    Image:	busybox:1.32
~/go/src/k8s.io/kubernetes (master) $ _output/bin/kubectl rollout history deploy/foo --revision=3 -o yaml | grep "image: busybox"
        image: busybox:1.32
```

I'm going to go ahead and close this issue since it has been fixed in kubectl 1.26.

/close
